standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/dataconnector/DataConnectorProviderFactory.java

```diff
@@ -18,29 +18,52 @@
 
 package org.apache.hadoop.hive.metastore.dataconnector;
 
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.RemovalCause;
+import com.github.benmanes.caffeine.cache.RemovalListener;
+import java.util.concurrent.TimeUnit;
 import org.apache.hadoop.hive.metastore.IHMSHandler;
 import org.apache.hadoop.hive.metastore.api.DataConnector;
 import org.apache.hadoop.hive.metastore.api.Database;
 import org.apache.hadoop.hive.metastore.api.DatabaseType;
 import org.apache.hadoop.hive.metastore.api.MetaException;
 import org.apache.hadoop.hive.metastore.api.NoSuchObjectException;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.util.HashMap;
-import java.util.Map;
-
-import static org.apache.hadoop.hive.metastore.dataconnector.IDataConnectorProvider.*;
+import static org.apache.hadoop.hive.metastore.dataconnector.IDataConnectorProvider.DERBY_TYPE;
+import static org.apache.hadoop.hive.metastore.dataconnector.IDataConnectorProvider.MSSQL_TYPE;
+import static org.apache.hadoop.hive.metastore.dataconnector.IDataConnectorProvider.MYSQL_TYPE;
+import static org.apache.hadoop.hive.metastore.dataconnector.IDataConnectorProvider.ORACLE_TYPE;
+import static org.apache.hadoop.hive.metastore.dataconnector.IDataConnectorProvider.POSTGRES_TYPE;
 
 public class DataConnectorProviderFactory {
-  Logger LOG = LoggerFactory.getLogger(DataConnectorProviderFactory.class);
+  static final Logger LOG = LoggerFactory.getLogger(DataConnectorProviderFactory.class);
 
-  private static Map<String, IDataConnectorProvider> cache = null;
+  private static Cache<String, IDataConnectorProvider> dataConnectorCache = null;
   private static DataConnectorProviderFactory singleton = null;
   private static IHMSHandler hmsHandler = null;
 
+  private static class CacheRemoveListener implements RemovalListener<String, IDataConnectorProvider> {
+    @Override
+    public void onRemoval(@Nullable String dcName, @Nullable IDataConnectorProvider dataConnectorProvider,
+                          @NonNull RemovalCause cause) {
+      try {
+        dataConnectorProvider.close();
+      } catch (Exception e) {
+        LOG.warn("Exception when closing dataconnectorprovider: {} due to: {}" + dcName, e.getMessage());
+      }
+    }
+  }
+
   private DataConnectorProviderFactory(IHMSHandler hmsHandler) {
-    cache = new HashMap<String, IDataConnectorProvider>();
+    dataConnectorCache = Caffeine.newBuilder()
+        .removalListener(new CacheRemoveListener())
+        .maximumSize(100)
```

Comment:
Should we make this configurable? WDYT

Reply:
I think it is not necessary to make this configurable.  We usually don't restrict cache size, but to avoid the extrem case to influence metastore memory, I set the maximumsize to 100 and this is enough for data connector.
BTW, the cache expire is one hour, and after one hour, the cache will be updated.

Reply:
Makes sense, thanks